### ISSUE TITLE
Lengthen penultimate stressed vowels in italian language

### ISSUE THREAD

**Issue by Christianlm** [espeak-ng/espeak-ng#80]
Currently espeak-ng  lengthens penultimate stressed vowels, unless followed by two consonants.

In italian language is also necessary don't lengthen before these phonemes:

/@-/ this is used to generate r in the sequence /@-*/. .

/dz/ 

/dZ/ 

/ts/ 

/tS/ 

/S/ this phoneme in italian is based from s and c consonants. 

/l^/ 

/n^/ 

/ss/ this phoneme is for long s between a vowel, recently added in ph_italian for testing.

Also is necessary don't lengthen when is preceded or is followed by a vowel.

E.g: the word ciao should be translate [tS'ao] instead of [tS'a:o].

Could be fixed this from the espeak-ng program?

Thanks.

**Comment by ValdisVitolins** [comment]
I checked it with latest commit [bbd981e](https://github.com/espeak-ng/espeak-ng/commit/bbd981efb88d5a81fdc99f8ca2a08f18bb46ba9d) but it says:
```
espeak-ng -vit -X "chao"
Found: 'chao' [_^_es]   $onlys
Unpronouncable? 'chao'
 26	_) ch    [tS]
Translate 'chao'
  1	c        [k]
 22	ch       [tS]
 26	_) ch    [tS]
  1	a        [a]
  1	o        [o]
 (es)tS'ao(it)
```
Because there is line `chao	_^_es` in [it_list](https://github.com/espeak-ng/espeak-ng/blob/master/dictsource/it_list). (By the way, what exactly means `$onlys`? Only stress?)
Christian, if you could provide list of words and correct pronunciations to test changes, I could try help with this.

**Comment by Christianlm** [comment]
My  simple question is:
it's possible delete the lengthening of penultimate stressed vowels in 
the situations described above?

By the way, in my example I wrote the italian word 'ciao', and not the 
imported spanish word 'chao'.

As far I know, $onlys mean:
"As $only, except that a standard  plural ending is allowed."

Thanks.





Chris.

Valdis Vitolins, 19/01/2016 22:10:

**Comment by ValdisVitolins** [comment]
With my knowledge espeak has means only to lengthen vowels. So, you can make them shorter in following ways:
1. Redefine existing vowels in `phsource/ph_italian` shorter, or add new definitions with shorter phonemes.
2. Create conditions to lengthen vowels:
2a. with language rules in `dictsource/it_rules` to select shorter or longer phoneme from `phsource/ph_italian` file. Note that here you can simply add `:` to make phoneme longer, or you can select different phonemes from `phsource/ph_italian` (and subsequently `phsource/phonemes`) files.
2b. or with [conditions](/espeak-ng/espeak-ng/blob/master/docs/phontab.md#conditional-statements) in `phsource/ph_italian` file.

**Comment by Christianlm** [comment]
If I'm not mistaken, the instructions to lengthen stress vowels for italian language are located in src\libespeak-ng\dictionary.c below line 1468.
@Reece, Could you delete  these instructions without affecting other languages?

My idea is:

1. Delete the lengthen stressed vowels from espeak-ng program.

2. The length for stressed vowels and nucleus in clause can be set in tr_language.c.

3. In ph_italian there are different statements for stressed and unstressed vowels, including the length parameter.

4. When is needed we can use a double vowel in dict source.

If you could edit dictionary.c, push it in espeak-italian branch so I can test and make the necessary changes for italian language.
Thanks.

**Comment by reece** [comment]
I think you meant @rhdunn

**Comment by Christianlm** [comment]
Sorry... Yes I mean @rhdunn.

**Comment by ValdisVitolins** [comment]
Christian,
now I'm more familiar with espeak features, and I found  [Stress lenght](https://github.com/espeak-ng/espeak-ng/blob/master/docs/voices.md#stresslength), therefore you don't need to change espeak-ng executables.
Probably you need to add following line into `espeak-data/voices/roa/it` file:
```
stressLength 180 180 180 180 0 0 180 180
```
Note that length of all types of stressed syllables is set to equal length: **180** (relative units).
Probably you will find, that diminished syllables should be shorter though, because for Latvian language (where stressed syllable also "shouldn't" be longer as other) I set following:
```
stressLength 180 180 180 160 0 0 220 180
```
(note 160 units for unstressed words and 220 units for stressed syllables).

**Comment by Christianlm** [comment]
Yes Valdis, I know this. I've set these values on tr_language.cpp time ago for eSpeak 1.48.3 release.
The problem is that eSpeak put in each case the length indicator, this generates a strange prosodic pitch. E.G. in words like ‘moglie’ ‘foglia’ ‘vuoto’ etc, the vowel [o] is doubbled  by eSpeak.
I'm not a sympathizer  of develope language, just I learned the eSpeak features to insert my linguistic knowledge. For this, if someone could to look  the section in dictionary.c mentioned above is appreciated.
Thanks.

**Comment by ValdisVitolins** [comment]
Christian, probably you are talking file what now is [dictionary.c lines 1469..1475](https://github.com/espeak-ng/espeak-ng/blob/master/src/libespeak-ng/dictionary.c#L1469).
In my opinion this is just simply wrong approach, because:
1. As you mentioned for your words **moglie, foglia, vuoto** it just works incorrectly anyway,
2. It is not traceable. E.g. look at traced output of pronunciation:
```
$ ~/code/espeak-ng/src/espeak-ng -X -vit "moglie. foglia. vuoto"
Translate 'moglie'
  1	  350:	m [m]

  1	  382:	o [o]

 43	  218:	gl (i [l^]
 63	  220:	gli (A [l^]
  1	  195:	g [g]

  1	  119:	e [e]

Flags:  foglia   $alt
Translate 'foglia'
  1	  186:	f [f]

  1	  382:	o [o]

 43	  218:	gl (i [l^]
 63	  220:	gli (A [l^]
  1	  195:	g [g]

  1	   25:	a [a]
 60	   30:	@i) a (_S1q  [a]

Flags:  fogli   $alt
Translate 'vuoto'
  1	  601:	v [v]

  1	  581:	u [u]

  1	  382:	o [o]

  1	  560:	t [t]

  1	  382:	o [o]
 58	  391:	@C) o (_S1q  [o/]

Found: 'vuot' [vU'Ot]  
 m'o:l^e f'O:l^a vU'O:to
```
as you can see, lengthening character `:` just pops out as devil without any trace, which is very hard to maintain.
@rhdunn I think, lines 1469..1475 should be just removed from `dictionary.c` (or `LOPT_IT_LENGTHEN` should be set to **1** somewhere) and lengthening should be described in `it_rules` file.

**Comment by rhdunn** [comment]
@ValdisVitolins Agreed. I would ultimately like the code to not have any special cases for languages as this makes it harder to maintain the language support and debug what is going on. Where needed, functionality should be provided in a generic way.

In this case, the logic should be encoded somewhere in the it_rules file. @Christianlm Is it possible for you to write the long vs short penultimate vowels in it_rules? Is there anything that espeak-ng is lacking that prevents you from doing this?

I would also like to move the tr_languages.c logic into the voice files.

**Comment by Christianlm** [comment]
@ValdisVitolins thanks very much. Can you try if after removing those lines eSpeak-ng don't length the penultimate vowels?
Deleting these lines don't affects oter languages?

@rhdunn in it_rules I can only whrite short vs long. When I whrite a vowel without leng indicator, espeak-ng put in any case the length indicator when the vowel is stressed, is penultimate and is not followed by two consonants.

I agree to move the tr_languages.c logic into the voice files, but in this case more documentations is needed for the languages maintainer.
Thanks.

**Comment by ValdisVitolins** [comment]
I commented out mentioned lines in `dictionary.c` and then did this in `it_rules` file (just example for letter a, please suggest necessary fixes):
```
.L20 a à e i o u     // vowels shouldn't à just be translated to a at beginning?
.L21 dz c ch g sh ss // shortening consonants (vowel is short before these, need actual list)

.group a
...
        @) a (L21+++       a  // a is short before these
        @) a (CL20_        a: // lengthen penultimate vowel
        @) a (CCL20_ 
```
The test:
```
$ esX -vit "pavane. pavace."
Translate 'pavane'
  1	p        [p]

  1	a        [a]

  1	v        [v]

  1	a        [a]
 62	@) a (CL20_ [a:]

  1	n        [n]

  1	e        [e]
 58	@C) e (_S1q  [e/]

Translate 'pavace'
  1	p        [p]

  1	a        [a]

  1	v        [v]

 64	C) a (cY_ ['a]
  1	a        [a]
100	@) a (L21+++ [a]
 62	@) a (CL20_ [a:]

  1	c        [k]
 21	c (Y     [tS]

  1	e        [e]
 58	@C) e (_S1q  [e/]

 pav'a:ne pav'atSe
```
@Christianlm Is this approach ok?
@rhdunn because it has to be added for each vowel and its spelling variants it actually will be many groups of similar rules for each letter. To make rules shorter, it would be nice, if one could make rule  translation with rules (issue #121) and syllables in post part of rules (issue #126), then it could be just one group, like this:
```
.translate
@) a (@_ a: // make penultimate syllable longer
@) e (@_ e:
@) i (@_ i:
@) o (@_ o:
@) u (@_ u:
```

**Comment by Christianlm** [comment]
Good idea, thanks. Just some comments:
we use "A" upercase to indicate any vowels, isn't  needed .L20 group.
the approach is ok, probably is more convenient create a group for consonants after a long vowel.
Thanks so much for your help, now I test and push in repository the last improvements.

**Comment by Christianlm** [comment]
Another comment.
The rules suggested
```
       @) a (CL20_        a: // lengthen penultimate vowel
```

is not correct for italian language cause lengthen the vowel  when is unstressed too.
In some case the primary stressed syllable in a word can be  the first syllable, last syllable etc...
Anyway, your suggestion is still valid, I must investigate more.
The most important thing for me is I can delete the lines in Dictionary.c without problem.
Thanks.

**Comment by ValdisVitolins** [comment]
Christianlm, I tested, how exactly it works and probably best way how to make penultimate syllable longer is following:

```
...
group a
a (@ a:
...
group e
e (@ e:
...
group i
i (@ i:
...
group o
o (@ o:
...
group u
u (@ u:
...
```
If necessary, you can add one or more pluses after `@` to make these rules stronger than others.
